Add optional argument to split

Jason2605 opened the pull request:
<!---- This is the PR Template !-->

<!-- Make sure to follow each step so that your PR is explained and easy to read !-->

<!-- This will allow maintainers and other potential contributors to understand the changes being carried out !-->

<!--- Thanks for considering that !-->

### Well detailed description of the change :

This PR adds a new optional argument to `str.split()` which determines the amount of times a string will actually be broken up by the delimiter. This allows us to keep the delimiter within the string after a certain amount of occurrences have already been found.

E.g
```
"Dictu is awesome!".split(" ", 0); // ['Dictu is awesome!']
"Dictu is awesome!".split(" ", 1); // ['Dictu', 'is awesome!']
"Dictu is awesome!".split(" ", 2); // ['Dictu', 'is', 'awesome!']
```

### Type of change:

<!-- Please select relevant options -->

<!-- Add an x in [ ] if true !-->

<!-- Delete options that aren't relevant!-->


- [ ] Bug fix

- [x] New feature

- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#

### Housekeeping

<!-- If adding a new test file, remember to include it in the relevant import file -->
- [x] Tests have been updated to reflect the changes done within this PR (if applicable).

- [x] Documentation has been updated to reflect the changes done within this PR (if applicable).

### Note
This may come in useful with regards to #473

briandowns commented:
I'm curious about your thoughts on this:

Keep basic functionality and expectations the same.

```cs
"Dictu is awesome!".split(" "); // ['Dictu', 'is', 'awesome!']
```

Add another method, like Go, to indicate the number of times to split.

```cs
"Dictu is awesome!".splitN(" ", 1); // ['Dictu', 'is awesome!']
"Dictu is awesome!".splitN(" ", 2); // ['Dictu', 'is', 'awesome!']
```

or

Have value indicating all instances of the seperator.

```cs
"Dictu is awesome!".split(" ", -1); // ['Dictu', 'is', 'awesome!']
```

Jason2605 commented:
Hey! Thanks for the comments!

> Add another method, like Go, to indicate the number of times to split.

I think another method will probably end up with a lot of code duplication or macro abuse which are never fun to deal with (neither are strings in C at the best of times ;( )

`.split(delim, occur)` is also how it is in Python 😄 

> Have value indicating all instances of the separator.

This is something I don't mind too much about. Checking in Python a negative number seems to make it ignore the count and just split by all occurrences of the delimiter (like how you propose with -1). The current implementation treats negative numbers as not wanting to split (so essentially the same as passing 0). Probably a good idea as it means you can have a negative number as a default rather than having to have two flows in user code with:

```
// Flow 1
var count = cond ? 3 : -1;
".........".split(",", count);

// Flow 2
if (cond) {
    ".........".split(",");
} else {
    ".........".split(",", 3);
}
```

Jason2605 commented:
@briandowns I've updated the PR to make negative numbers split all occurrences now, thanks for the thought!

briandowns commented:
LGTM!
